text/1011-object-level-security.rst

```diff
@@ -0,0 +1,320 @@
+      property private -> bool;
+
+      # ... but restrict access to private posts to owner only
+      # regardless of what permissions were granted in parent types
+      access policy private_owner_only
+        when (.private)
+        restrict all using (.owner = global current_user);
+    }
+
+
+
+ALTER ACCESS POLICY
+-------------------
+
+Alter the definition of an access control policy.
+
+Required capabilities: DDL.
+
+Synopsis::
+
+    ALTER OBJECT TYPE <type-name> "{"
+        ALTER ACCESS POLICY <name>
+            [ WHEN <condition> ]
+            { PERMIT | RESTRICT }
+            { ALL | READ | WRITE | DELETE } [ ... ]
+            USING (<expr>)
+        [ "{" <subcommand>; [...] "}" ];
+    "}"
+
+    # where <subcommand> is one of
+
+      CREATE ANNOTATION <annotation-name> := <value>
+      ALTER ANNOTATION <annotation-name> := <value>
+      DROP ANNOTATION <annotation-name>
+
+
+DROP ACCESS POLICY
+------------------
+
+Remove an access control policy.
+
+Required capabilities: DDL.
+
+Synopsis::
+
+    ALTER OBJECT TYPE <type-name> "{"
+        DROP ACCESS POLICY <name>;
+    "}"
+
+
+Interaction with globals
+========================
+
+Access policies are especially powerful when combined with RFC 1010
+globals, because then data visibility can be globally adjusted with a single
+``SET GLOBAL`` statement, which is very useful for authenticated/authorized
+data access control.
+
+Example::
+
+    global user_id -> uuid;
+
+    abstract object type Owned {
+      required link owner -> User;
+
+      access policy owner_only
+        permit all (.owner.id = global user_id)
+    }
+
+    object type Purchase extending Owned;
+
+    ...
+
+    set global user_id := <uuid-1>;
+    select count(Purchase);
+    # 9
+    set global user_id := <uuid-2>
+    select count(Purchase);
+    # 1
+
+
+Bypassing policies
+==================
+
+A superuser can bypass the execution of query rewrite policies by setting
+the ``apply_access_policies`` session configuration setting to ``false``.
+
+
+Mandatory Role-based Access Control (RBAC)
+==========================================
+
+Coupled with the role-based permission system (discussed in a future RFC),
+object-level security provides reliable mandatory RBAC, where an
+``access policy`` is protected by role permissions and cannot be disabled
+by unauthorized users.
+
+
+Introspection
+=============
+
+Policies can be introspected via a new ``schema::AccessPolicy`` in the
+introspection schema that is linked from ``schema::ObjectType`` via the new
+``access_policies`` link.  The ``schema::AccessPolicy`` is exposed as follows::
+
+    type schema::AccessPolicy extending schema::AnnotationSubject {
+      multi property access_kind -> schema::AccessKind;
+      property condition -> std::str;
+      required property action -> schema::AccessPolicyAction;
+      required property expr -> std::str;
+    };
+
+
+Implementation considerations
+=============================
```

Comment:
One implementation consideration is how to apply access policies in subtypes. I can think of two approaches:
 1. Instead of using views for types with access rules, fully expand it out into a union of all the subtypes, each with appropriate filtering rules
 2. Keep using views, and use type filtering to do the appropriate filters on subtypes

Reply:
I was thinking that the most reliable solution would be to actually _always_ bake all access policies into views themselves.

Reply:
How would they access global variables, then?

Reply:
Hm. Good point, it slipped from my mind that we're moving away from the temp table approach altogether.  Full expansion would be the best option from the standpoint of query quality, though at the cost of query text size.  A set-returning function instead of a view would be another option, but it's probably too slow and might be too opaque for the optimizer.